packagedb/find_source_repo.py

```diff
@@ -0,0 +1,214 @@
+    # remove .git from the end of the URL
+    if source_url.endswith(".git"):
+        source_url = source_url[:-4]
+    if source_url.startswith("git://"):
+        # remove git:// from the beginning of the URL
+        source_url = source_url[6:]
+        if ":" in source_url:
+            # convert : to /
+            source_url = source_url.replace(":", "/")
+        source_url = f"https://{source_url}"
+    urlparse_result = urlparse(source_url)
+    if not len(urlparse_result.path.split("/")) > 2:
+        return
+    namespace = urlparse_result.path.split("/")[1]
+    name = urlparse_result.path.split("/")[2]
+    if not name:
+        return
+    for url_hint in url_hints:
+        if url_hint in urlparse_result.netloc:
+            return PackageURL(
+                type=url_hint,
+                namespace=namespace,
+                name=name,
+            )
+
+
+def get_url_from_package_resources(package):
+    """
+    Return the URL of the source repository of a package
+    """
+    urls = []
+    for resource in package.resources.filter(is_key_file=True):
+        urls = [url["url"] for url in resource.urls]
+    return get_git_repo_url(urls=urls)
+
+
+def get_url_from_package_data(package):
+    """
+    Return the URL of the source repository of a package
+    """
+    # TODO: Use the source package url
+    # TODO: If the package is already a repo package then don't do anything
+    # TODO: Search for URLs in description, qualifiers, download_url, notice_text, extracted_license_statement.
+    return get_git_repo_url(
+        urls=[
+            package.vcs_url,
+            package.homepage_url,
+            package.code_view_url,
+            package.bug_tracking_url,
+            package.repository_homepage_url,
+            package.repository_download_url,
+        ]
+    )
+
+
+def get_git_repo_url(urls):
+    """
+    Return the first URL that contains a git repo URL
+    """
+    # TODO: Refine this
+    url_hints = [
+        "github",
+        "gitlab",
+        "bitbucket",
+    ]
+    for url in urls:
+        if url and any(url_hint in url for url_hint in url_hints):
+            return url
+
+
+def get_tags_and_commits(source_purl):
+    """
+    Yield tuples of (tag, commit), given a source_purl PackageURL
+    """
+    try:
+        repo_url = purl2url(str(source_purl))
+        resp = requests.get(repo_url)
+        if resp.status_code != 200:
+            return
+        output = subprocess.getoutput(f"git ls-remote {repo_url}")
+        yield from get_tags_and_commits_from_git_output(output)
+    except Exception as e:
+        logger.error(f"Error getting tags and commits for {source_purl}: {e}")
+
+
+def get_tags_and_commits_from_git_output(git_ls_remote):
+    """
+    Yield tuples of (tag, commit), given a git ls-remote output
+    """
+    for line in git_ls_remote.split("\n"):
+        line_segments = line.split("\t")
+        if len(line_segments) > 1 and line_segments[1].startswith("refs/tags/"):
```

Comment:
It would be nice to have a comment that shows an example line, so it helps the reader visualize what segments of it you're using in your code.